lib/spack/spack/util/compression.py

```diff
@@ -23,20 +30,153 @@ def allowed_archive(path):
     return any(path.endswith(t) for t in ALLOWED_ARCHIVE_TYPES)
 
 
+def try_exec(executor):
+    try:
+        return which(executor, required=True)
+    except CommandNotFoundError:
+        pass
+
+
+@contextmanager
+def exploding_archive_catch(name):
+    unpack_dir = 'spack_unpack_stage'
+    catch_dir = os.path.join(os.getcwd(), unpack_dir)
+    mkdirp(catch_dir)
+    orig_dir = os.getcwd()
+    os.chdir(catch_dir)
+    try:
+        yield
+    finally:
+        # catch an exploding archive
+        os.chdir(orig_dir)
+        unpacked_contents = os.listdir(catch_dir)
+        metadata = [x for x in unpacked_contents if x.startswith('.')]
+        unpacked_contents_set = set(unpacked_contents) - set(metadata)
+
+        if len(unpacked_contents_set) > 1:
+            os.rename(catch_dir, os.path.join(orig_dir, name))
+        elif unpacked_contents_set:
+            for f in unpacked_contents + metadata:
+                f_path = os.path.join(catch_dir, f)
+                if os.path.isdir(f_path):
+                    copy_tree(os.path.join(catch_dir, f),
+                              os.path.join(orig_dir, f))
+                else:
+                    copy(os.path.join(catch_dir, f),
+                         os.path.join(orig_dir, f))
+            shutil.rmtree(catch_dir)
+
+
+def derive_fallback_executable(fallback):
+    for strategy in fallback:
+        test_exec = strategy
+        islist = False
+        if not isinstance(strategy, string_types):
+            test_exec = strategy[0]
+            islist = True
+        executor = try_exec(test_exec)
+        if executor:
+            args = []
+            if islist:
+                args = strategy[1:]
+            return executor, args
+    raise CommandNotFoundError("Unable to find system fallback unpacking utility")
+
+
+def _system_fallback(archive_file):
+    import spack.config as sconf
+    outfile = os.path.basename(archive_file)
+    fallback = sconf.config.get('config:fallback_decompression_strategy')
+    if isinstance(fallback, string_types):
+        fallback = [fallback]
+    extractor, args = derive_fallback_executable(fallback)
+    if args:
+        [extractor.add_default_arg(arg) for arg in args]
+    extractor(archive_file)
+    return outfile
+
+
+def _untar(archive_file):
+    """ Untar archive. Prefer native Python `tarfile`
+    but fall back to system utility failing
+    to find the native Python module (tar on Unix).
+    Filters archives through native support gzip and xz
+    compression formats.
+
+    Args:
+        archive_file (str): absolute path to the archive to be extracted.
+        Can be one of .tar(.[gz|bz2|xz|Z]) or .(tgz|tbz|tbz2|txz).
+    """
+    _, ext = os.path.splitext(archive_file)
+    outfile = os.path.basename(archive_file.strip(ext))
+    remnant = os.path.join(os.getcwd(), archive_file)
+
+    with exploding_archive_catch(outfile):
+        try:
+            import tarfile
```

Comment:
(question) is it possible that the decompression modules might have similar problems as `zipfile` (i.e. could this run into the same issue as https://github.com/spack/spack/pull/30407)?

Reply:
Unsure, will test for myself to determine on unix/windows.

Reply:
gunzip should prefer Python support. Here we've just introduced backup support for the command line utility, so I'll verify that as well. Windows does not natively bundle a number of system tools needed to extract each of these extension types. This issue (particularly with bz2) is the initial motivator for this PR. We now have 7zip as a spack package, so between that and the native tar we cover most scenarios, however 7zip source code is distributed as a `.7z`, which can only be extracted by 7zip, and `.tar.xz` which Windows tar does not natively support, so we must rely on Python initially. I'm also of the opinion that for improved cross platform support we should rely more on the Python based options wherever available (Unzip is an obvious exception).